File: frontend/src/api/mistake.ts
/**
 * 错题本 API 客户端
 * 处理错题的增删改查、复习等操作
 */
import axios from 'axios'

const api = axios.create({
  baseURL: '/api/v1',
  headers: {
    'Content-Type': 'application/json'
  }
})

// 请求拦截器：添加 token
api.interceptors.request.use(
  (config: any) => {
    const token = localStorage.getItem('access_token')
    if (token) {
      config.headers.Authorization = `Bearer ${token}`
    }
    return config
  },
  (error: any) => {
    return Promise.reject(error)
  }
)

// 响应拦截器：处理错误
api.interceptors.response.use(
  (response: any) => response.data,
  (error: any) => {
    if (error.response?.status === 401) {
      // 清除 token，跳转登录
      localStorage.removeItem('access_token')
      localStorage.removeItem('refresh_token')
      window.location.href = '/login'
    }
    return Promise.reject(error)
  }
)

/**
 * 错题类型枚举
 */
export enum MistakeType {
  GRAMMAR = 'grammar',
  VOCABULARY = 'vocabulary',
  READING = 'reading',
  LISTENING = 'listening',
  WRITING = 'writing',
  SPEAKING = 'speaking',
  PRONUNCIATION = 'pronunciation',
  COMPREHENSION = 'comprehension'
}

/**
 * 错题状态枚举
 */
export enum MistakeStatus {
  PENDING = 'pending',
  REVIEWING = 'reviewing',
  MASTERED = 'mastered',
  IGNORED = 'ignored'
}

/**
 * 错题类型定义
 */
export interface Mistake {
  id: string
  student_id: string
  practice_id?: string
  content_id?: string
  question: string
  wrong_answer: string
  correct_answer: string
  mistake_type: MistakeType
  status: MistakeStatus
  explanation?: string
  knowledge_points: string[]
  difficulty_level?: string
  topic?: string
  mistake_count: number
  review_count: number
  last_reviewed_at?: string
  first_mistaken_at?: string
  last_mistaken_at?: string
  ai_suggestion?: string
  ai_analysis?: Record<string, unknown>
  extra_metadata?: Record<string, unknown>
  mastery_level: number
  needs_review: boolean
  is_mastered: boolean
  created_at: string
  updated_at: string
}

/**
 * 创建错题请求
 */
export interface CreateMistakeRequest {
  question: string
  wrong_answer: string
  correct_answer: string
  mistake_type: MistakeType
  practice_id?: string
  content_id?: string
  explanation?: string
  knowledge_points?: string[]
  difficulty_level?: string
  topic?: string
  extra_metadata?: Record<string, unknown>
}

/**
 * 错题列表响应
 */
export interface MistakeListResponse {
  total: number
  limit: number
  offset: number
  mistakes: Mistake[]
}

/**
 * 错题统计数据
 */
export interface MistakeStatistics {
  student_id: string
  total_mistakes: number
  by_status: Record<string, number>
  by_type: Record<string, number>
  by_topic: Record<string, number>
  mastery_rate: number
  need_review_count: number
  recent_mistakes_count: number
  frequent_mistakes_count: number
}

/**
 * 复习计划
 */
export interface ReviewPlan {
  student_id: string
  urgent: Array<{ id: string; question: string; type: string }>
  today: Array<{ id: string; question: string; type: string }>
  week: Array<{ id: string; question: string; type: string }>
  knowledge_points: Array<[string, number]>
  total_pending: number
}

/**
 * 重做请求
 */
export interface RetryMistakeRequest {
  user_answer: string
  is_correct: boolean
}

/**
 * 重做响应
 */
export interface RetryMistakeResponse {
  mistake_id: string
  mastered: boolean
  review_count: number
  mistake_count: number
  status: MistakeStatus
  message: string
}

/**
 * 错题本 API
 */
export const mistakeApi = {
  /**
   * 创建错题记录
   */
  async createMistake(data: CreateMistakeRequest): Promise<Mistake> {
    return api.post('/mistakes/', data)
  },

  /**
   * 从练习记录中收集错题
   */
  async collectFromPractice(practiceId: string): Promise<{
    practice_id: string
    collected_count: number
    mistakes: Array<{
      id: string
      question: string
      mistake_type: string
      status: string
    }>
    message: string
  }> {
    return api.post(`/mistakes/collect/${practiceId}`)
  },

  /**
   * 获取错题列表
   */
  async getMyMistakes(params?: {
    status?: MistakeStatus
    mistake_type?: MistakeType
    topic?: string
    knowledge_point?: string
    needs_ai_analysis?: boolean
    limit?: number
    offset?: number
  }): Promise<MistakeListResponse> {
    return api.get('/mistakes/me', { params })
  },

  /**
   * 获取错题详情
   */
  async getMistake(mistakeId: string): Promise<Mistake> {
    return api.get(`/mistakes/${mistakeId}`)
  },

  /**
   * 更新错题状态
   */
  async updateMistakeStatus(
    mistakeId: string,
    status: MistakeStatus
  ): Promise<{ id: string; status: string; message: string }> {
    return api.put(`/mistakes/${mistakeId}/status`, { status })
  },

  /**
   * 重做错题
   */
  async retryMistake(
    mistakeId: string,
    data: RetryMistakeRequest
  ): Promise<RetryMistakeResponse> {
    return api.post(`/mistakes/${mistakeId}/retry`, data)
  },

  /**
   * 删除错题
   */
  async deleteMistake(mistakeId: string): Promise<void> {
    return api.delete(`/mistakes/${mistakeId}`)
  },

  /**
   * 获取错题统计数据
   */
  async getStatistics(): Promise<MistakeStatistics> {
    return api.get('/mistakes/me/statistics')
  },

  /**
   * 获取复习计划
   */
  async getReviewPlan(limit = 20): Promise<ReviewPlan> {
    return api.get('/mistakes/me/review-plan', { params: { limit } })
  },

  /**
   * 对单个错题进行AI分析
   */
  async analyzeMistake(mistakeId: string): Promise<{
    mistake_id: string
    analysis: {
      mistake_category: string
      severity: string
      explanation: string
      correct_approach: string
      knowledge_points: string[]
      recommendations: Array<{
        priority: number
        category: string
        title: string
        description: string
        resources?: string[]
        practice_exercises?: string[]
        estimated_time?: string
      }>
      review_plan: {
        review_frequency: string
        next_review_days: number[]
        mastery_criteria: string
        review_method: string
      }
      encouragement: string
    }
    message: string
  }> {
    return api.post(`/mistakes/${mistakeId}/analyze`)
  },

  /**
   * 批量AI分析错题
   */
  async batchAnalyzeMistakes(limit = 10): Promise<{
    analyzed_count: number
    summary: string
    common_patterns: string[]
    overall_recommendations: string[]
    priority_topics: string[]
    remaining_count: number
    message: string
  }> {
    return api.post('/mistakes/batch-analyze', null, { params: { limit } })
  },

  /**
   * 导出错题本
   */
  async exportMistakes(params: {
    format_type: 'markdown' | 'pdf' | 'word'
    status_filter?: string
    type_filter?: string
    topic_filter?: string
    knowledge_point_filter?: string
  }): Promise<void> {
    // 构建查询参数
    const queryParams = new URLSearchParams()
    queryParams.append('format_type', params.format_type)
    if (params.status_filter) queryParams.append('status_filter', params.status_filter)
    if (params.type_filter) queryParams.append('type_filter', params.type_filter)
    if (params.topic_filter) queryParams.append('topic_filter', params.topic_filter)
    if (params.knowledge_point_filter) queryParams.append('knowledge_point_filter', params.knowledge_point_filter)

    // 创建请求
    const token = localStorage.getItem('access_token')
    const url = `/api/v1/mistakes/export?${queryParams.toString()}`

    // 发起请求并下载文件
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || '导出失败')
    }

    // 从响应头获取文件名
    const contentDisposition = response.headers.get('Content-Disposition')
    let filename = '错题本导出'
    if (contentDisposition) {
      const filenameMatch = contentDisposition.match(/filename\*=UTF-8''(.+)/)
      if (filenameMatch && filenameMatch[1]) {
        filename = decodeURIComponent(filenameMatch[1])
      }
    }

    // 获取文件内容
    const blob = await response.blob()

    // 创建下载链接
    const downloadUrl = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = downloadUrl
    link.download = filename
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    window.URL.revokeObjectURL(downloadUrl)
  },

  /**
   * 导出单个错题
   */
  async exportSingleMistake(
    mistakeId: string,
    format_type: 'markdown' | 'pdf' | 'word'
  ): Promise<void> {
    // 构建查询参数
    const queryParams = new URLSearchParams()
    queryParams.append('format_type', format_type)

    // 创建请求
    const token = localStorage.getItem('access_token')
    const url = `/api/v1/mistakes/${mistakeId}/export?${queryParams.toString()}`

    // 发起请求并下载文件
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Authorization': `Bearer ${token}`,
      },
    })

    if (!response.ok) {
      const error = await response.json()
      throw new Error(error.detail || '导出失败')
    }

    // 从响应头获取文件名
    const contentDisposition = response.headers.get('Content-Disposition')
    let filename = '错题导出'
    if (contentDisposition) {
      const filenameMatch = contentDisposition.match(/filename\*=UTF-8''(.+)/)
      if (filenameMatch && filenameMatch[1]) {
        filename = decodeURIComponent(filenameMatch[1])
      }
    }

    // 获取文件内容
    const blob = await response.blob()

    // 创建下载链接
    const downloadUrl = window.URL.createObjectURL(blob)
    const link = document.createElement('a')
    link.href = downloadUrl
    link.download = filename
    document.body.appendChild(link)
    link.click()
    document.body.removeChild(link)
    window.URL.revokeObjectURL(downloadUrl)
  }
}

export default mistakeApi

File: backend/app/models/mistake.py
"""
错题本模型 - AI英语教学系统
记录学生在练习中的错误题目，支持错题复习和巩固
"""
import uuid
from datetime import datetime
from enum import Enum as PyEnum
from typing import TYPE_CHECKING, Optional

from sqlalchemy import Boolean, DateTime, Enum, ForeignKey, Integer, JSON, String, Text, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base

if TYPE_CHECKING:
    from app.models.student import Student
    from app.models.content import Content
    from app.models.practice import Practice


class MistakeStatus(str, PyEnum):
    """错题状态枚举"""
    PENDING = "pending"           # 待复习
    REVIEWING = "reviewing"       # 复习中
    MASTERED = "mastered"         # 已掌握
    IGNORED = "ignored"           # 已忽略


class MistakeType(str, PyEnum):
    """错题类型枚举"""
    GRAMMAR = "grammar"           # 语法错误
    VOCABULARY = "vocabulary"     # 词汇错误
    READING = "reading"           # 阅读理解错误
    LISTENING = "listening"       # 听力错误
    WRITING = "writing"           # 写作错误
    SPEAKING = "speaking"         # 口语错误
    PRONUNCIATION = "pronunciation"  # 发音错误
    COMPREHENSION = "comprehension"  # 理解错误


class Mistake(Base):
    """
    错题本模型

    记录学生在练习中的错误题目，支持以下功能：
    1. 自动收集错题（从练习记录中提取）
    2. 按知识点分类整理
    3. AI生成解析和建议
    4. 支持错题重做
    5. 追踪掌握程度
    """

    __tablename__ = "mistakes"

    # 主键
    id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        primary_key=True,
        default=uuid.uuid4,
        index=True
    )

    # 关联的学生ID
    student_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("students.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )

    # 关联的练习记录ID（可选，用于追溯）
    practice_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("practices.id", ondelete="SET NULL"),
        nullable=True,
        index=True
    )

    # 关联的内容ID（可选）
    content_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("contents.id", ondelete="SET NULL"),
        nullable=True,
        index=True
    )

    # 错题类型
    mistake_type: Mapped[str] = mapped_column(
        String(50),
        nullable=False,
        index=True
    )

    # 错题状态
    status: Mapped[str] = mapped_column(
        String(50),
        default=MistakeStatus.PENDING.value,
        index=True
    )

    # 题目内容
    question: Mapped[str] = mapped_column(
        Text,
        nullable=False
    )

    # 学生错误答案
    wrong_answer: Mapped[str] = mapped_column(
        Text,
        nullable=False
    )

    # 正确答案
    correct_answer: Mapped[str] = mapped_column(
        Text,
        nullable=False
    )

    # 错题解析（AI生成或教师添加）
    explanation: Mapped[Optional[str]] = mapped_column(Text)

    # 知识点标签（JSON数组）
    knowledge_points: Mapped[Optional[list]] = mapped_column(
        JSON,
        default=list
    )

    # 难度等级
    difficulty_level: Mapped[Optional[str]] = mapped_column(
        String(50),
        nullable=True
    )

    # 主题分类
    topic: Mapped[Optional[str]] = mapped_column(
        String(100),
        nullable=True,
        index=True  # 按知识点分类查询常用
    )

    # 错误次数
    mistake_count: Mapped[int] = mapped_column(
        Integer,
        default=1
    )

    # 复习次数
    review_count: Mapped[int] = mapped_column(
        Integer,
        default=0
    )

    # 最后复习时间
    last_reviewed_at: Mapped[Optional[datetime]] = mapped_column(
        DateTime,
        nullable=True
    )

    # AI生成的学习建议
    ai_suggestion: Mapped[Optional[str]] = mapped_column(Text)

    # 是否需要AI深度分析
    needs_ai_analysis: Mapped[bool] = mapped_column(
        Boolean,
        default=True
    )

    # AI分析结果（JSON格式）
    ai_analysis: Mapped[Optional[dict]] = mapped_column(
        JSON,
        default=dict
    )

    # 扩展元数据（JSON格式，存储题目选项、上下文等）
    extra_metadata: Mapped[Optional[dict]] = mapped_column(
        JSON,
        default=dict
    )

    # 首次错误时间
    first_mistaken_at: Mapped[datetime] = mapped_column(
        DateTime,
        server_default=func.now(),
        nullable=False
    )

    # 最后更新时间
    last_mistaken_at: Mapped[datetime] = mapped_column(
        DateTime,
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False
    )

    # 时间戳
    created_at: Mapped[datetime] = mapped_column(
        DateTime,
        server_default=func.now(),
        nullable=False
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime,
        server_default=func.now(),
        onupdate=func.now(),
        nullable=False
    )

    # 关系 - 关联的学生
    student: Mapped["Student"] = relationship(
        "Student",
        back_populates="mistakes",
        foreign_keys=[student_id]
    )

    # 关系 - 关联的练习记录
    practice: Mapped[Optional["Practice"]] = relationship(
        "Practice",
        foreign_keys=[practice_id]
    )

    # 关系 - 关联的内容
    content: Mapped[Optional["Content"]] = relationship(
        "Content",
        foreign_keys=[content_id]
    )

    def __repr__(self) -> str:
        return (
            f"<Mistake(id={self.id}, student_id={self.student_id}, "
            f"type={self.mistake_type}, status={self.status}, "
            f"mistake_count={self.mistake_count})>"
        )

    @property
    def needs_review(self) -> bool:
        """是否需要复习（状态为待复习或复习中）"""
        return self.status in [
            MistakeStatus.PENDING.value,
            MistakeStatus.REVIEWING.value
        ]

    @property
    def is_mastered(self) -> bool:
        """是否已掌握"""
        return self.status == MistakeStatus.MASTERED.value

    @property
    def mastery_level(self) -> float:
        """
        掌握程度 (0-1)

        基于以下因素计算：
        - 复习次数
        - 错误次数趋势
        - 当前状态
        """
        if self.is_mastered:
            return 1.0

        # 基础分数：复习次数贡献
        base_score = min(self.review_count * 0.2, 0.6)

        # 状态加成
        if self.status == MistakeStatus.REVIEWING.value:
            base_score += 0.2

        return min(base_score, 1.0)
